### PR TITLE
Hydro Pybindings and Test Cases

### PR DESCRIPTION
**Pull request by aasthabagree21**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/PointwiseFunctions/Hydro/Python/ComovingMagneticField.hpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+
+namespace py_bindings {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/ComovingMagneticField.hpp:10]: Should this be `hydro::py_bindings`? @nilsvu what's our convention here?

File: src/PointwiseFunctions/Hydro/Python/ComovingMagneticField.hpp [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+// NOLINTNEXTLINE(google-runtime-references)
+void bind_comovingMF(pybind11::module& m);
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/ComovingMagneticField.hpp]: `bind_comoving_magnetic_field` (this violates naming convention and MF is a bit of an unfortunate abbreviation that I don't think we should have in a public code 🙃 )

File: src/PointwiseFunctions/Hydro/Python/ComovingMagneticField.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include <string>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "PointwiseFunctions/Hydro/ComovingMagneticField.hpp"
+
+namespace py = pybind11;
+
+namespace py_bindings {
+template <typename DataType>
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/ComovingMagneticField.cpp:18]: put `_impl` into anonymous namespace

File: src/PointwiseFunctions/Hydro/Python/LorentzFactor.cpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+#include <string>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "PointwiseFunctions/Hydro/LorentzFactor.hpp"
+
+namespace py = pybind11;
+
+namespace py_bindings {
+template <typename DataType>
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/LorentzFactor.cpp:17]: put `_impl` into anonymous namespace

File: src/PointwiseFunctions/Hydro/Python/LorentzFactor.hpp [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+
+namespace py_bindings {
+// NOLINTNEXTLINE(google-runtime-references)
+void bind_lorentz(pybind11::module& m);
+
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/LorentzFactor.hpp:13]: probably need the same `NOLINTNEXTLINE` here

File: src/PointwiseFunctions/Hydro/Python/LowerSpatialFourVelocity.hpp [hunk 6]
```diff
@@ -0,0 +1,11 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+
+namespace py_bindings {
+// NOLINTNEXTLINE(google-runtime-references)
+void bind_lowerVel(pybind11::module& m);
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/LowerSpatialFourVelocity.hpp]: `bind_lower_spatial_four_velocity`

File: src/PointwiseFunctions/Hydro/Python/LowerSpatialFourVelocity.cpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+#include <pybind11/stl.h>
+#include <string>
+
+#include "PointwiseFunctions/Hydro/LowerSpatialFourVelocity.hpp"
+
+namespace py = pybind11;
+
+namespace py_bindings {
+void bind_lowerVel(py::module& m) {
+  m.def("LowerSpatialFourVelocityCompute",
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/LowerSpatialFourVelocity.cpp:16]: `lower_spatial_four_velocity` (you are binding a function, ultimately)

File: src/PointwiseFunctions/Hydro/Python/MassFlux.cpp [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+#include <string>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "PointwiseFunctions/Hydro/MassFlux.hpp"
+
+namespace py = pybind11;
+
+namespace py_bindings {
+template <typename DataType, size_t Dim, typename Frame>
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/MassFlux.cpp:17]: put `_impl` into anonymous namespace

File: src/PointwiseFunctions/Hydro/Python/MassFlux.hpp [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+
+namespace py_bindings {
+// NOLINTNEXTLINE(google-runtime-references)
+void bind_massFlux(pybind11::module& m);
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/MassFlux.hpp]: `bind_mass_flux`

File: src/PointwiseFunctions/Hydro/Python/MassWeightedFluidItems.hpp [hunk 10]
```diff
@@ -0,0 +1,11 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+
+namespace py_bindings {
+// NOLINTNEXTLINE(google-runtime-references)
+void bind_massWeighted(pybind11::module& m);
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/MassWeightedFluidItems.hpp]: `bind_mass_weighted`

File: src/PointwiseFunctions/Hydro/Python/SoundSpeedSquared.cpp [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "PointwiseFunctions/Hydro/EquationsOfState/EquationOfState.hpp"
+#include "PointwiseFunctions/Hydro/SoundSpeedSquared.hpp"
+
+namespace py = pybind11;
+
+namespace py_bindings {
+template <typename DataType, size_t ThermodynamicDim>
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/SoundSpeedSquared.cpp:18]: put `_impl` into anonymous namespace

File: src/PointwiseFunctions/Hydro/Python/SoundSpeedSquared.hpp [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+
+namespace py_bindings {
+// NOLINTNEXTLINE(google-runtime-references)
+void bind_soundSpeed(pybind11::module& m);
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/SoundSpeedSquared.hpp]: `bind_sound_speed`

File: src/PointwiseFunctions/Hydro/Python/SpecificEnthalpy.cpp [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "PointwiseFunctions/Hydro/SpecificEnthalpy.hpp"
+
+namespace py = pybind11;
+
+namespace py_bindings {
+template <typename DataType>
+void bind_specificEnthalpy_impl(py::module& m) {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/SpecificEnthalpy.cpp]: put `_impl` into anonymous namespace

File: src/PointwiseFunctions/Hydro/Python/SpecificEnthalpy.hpp [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+
+#pragma once
+
+#include <pybind11/pybind11.h>
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+
+namespace py_bindings {
+// NOLINTNEXTLINE(google-runtime-references)
+void bind_specificEnthalpy(pybind11::module& m);
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/Python/SpecificEnthalpy.hpp]: `bind_specific_enthalpy` (I will now stop adding these comments, just change all the functions to snake_case please :) )

File: src/PointwiseFunctions/Hydro/MassWeightedFluidItems.hpp [hunk 15]
```diff
@@ -60,6 +60,15 @@ struct MassWeightedCoords : db::SimpleTag {
 };
 }  // namespace Tags
 
+
+template <typename DataType, size_t Dim, typename Frame>
+void u_lower_t(const gsl::not_null<Scalar<DataType>*> result,
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/Hydro/MassWeightedFluidItems.hpp:65]: Add documentation?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Overall looks good. Here are some suggestions, we can also talk about the compiler error once you're back. There might be some python tests still missing, but also something we can discuss in terms of how best to do this strategically. You can squash any changes since most of mine are just renames :)

File: src/PointwiseFunctions/Hydro/Python/LorentzFactor.hpp (hunk 5 above)
Comment (aasthabagree21) [reply src/PointwiseFunctions/Hydro/Python/LorentzFactor.hpp:13]: Which file requires this line?

File: src/PointwiseFunctions/Hydro/MassWeightedFluidItems.hpp (hunk 15 above)
Comment (aasthabagree21) [reply src/PointwiseFunctions/Hydro/MassWeightedFluidItems.hpp:65]: Hi where should I add the documentation?

File: src/PointwiseFunctions/Hydro/Python/LorentzFactor.hpp (hunk 5 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/Hydro/Python/LorentzFactor.hpp:13]: I think this file, this line to deal with the `pybind11::module &` on the next line

File: src/PointwiseFunctions/Hydro/MassWeightedFluidItems.hpp (hunk 15 above)
Comment (nilsdeppe) [reply src/PointwiseFunctions/Hydro/MassWeightedFluidItems.hpp:65]: Just a few lines up, line 63.
```cpp
/// \brief DOCUMENTATION HERE
template <typename DataType, size_t Dim, typename Frame>
void u_lower_t(const gsl::not_null<Scalar<DataType>*> result,
```

File: src/PointwiseFunctions/Hydro/Python/Bindings.cpp [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+  py::module_::import("spectre.Spectral");
+  py_bindings::bind_comoving_magnetic_field(m);
+  py_bindings::bind_lorentz(m);
+  py_bindings::bind_lorentz_factor(m);
+  py_bindings::bind_lower_spatial_four_velocity(m);
+  py_bindings::bind_mass_flux(m);
+  py_bindings::bind_mass_weighted(m);
+  py_bindings::bind_sound_speed(m);
+  py_bindings::bind_specific_enthalpy(m);
+  py_bindings::bind_stress_energy(m);
```
Comment (nilsvu) [inline src/PointwiseFunctions/Hydro/Python/Bindings.cpp:33]: Please just put these all into one this file, like `PointwiseFunctions/GeneralRelativity/Python/Bindings.cpp`. That just simplifies the code by a lot.

You don't need to worry about templates on `Frame` or `DataType`. It's enough to fix those to `Frame::Inertial` (the default, so just delete it in most places) and `DataVector` for now.

File: src/PointwiseFunctions/Hydro/Python/ComovingMagneticField.hpp (hunk 1 above)
Comment (nilsvu) [reply src/PointwiseFunctions/Hydro/Python/ComovingMagneticField.hpp:10]: Yep, `hydro::pybindings`

File: tests/Unit/PointwiseFunctions/Hydro/Python/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,38 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+spectre_add_python_bindings_test(
+  "Unit.Hydro.ComovingMagneticField.Python"
+  Test_ComovingMagneticField.py
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/Hydro/Python/CMakeLists.txt:6]: Also these tests all in one file, `Test_Bindings.py`.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/Hydro/Python/Bindings.cpp (hunk 16 above)
Comment (kidder) [reply src/PointwiseFunctions/Hydro/Python/Bindings.cpp:33]: why?  @nilsdeppe and I told @aasthabagree21 not to have large monolithic files...

Comment (nilsvu) [reply src/PointwiseFunctions/Hydro/Python/Bindings.cpp:33]: We should communicate better so @aasthabagree21 is not in the middle of this. The reason is lots of unnecessary boilerplate glue code. For the GR pointwise bindings we have recently moved to a list of pybindings in GR/Python/Bindings.cpp and that worked very well, it made it very easy for Geoffrey's students to add all those bindings. I think @nilsdeppe and I discussed to try that at the time, spawned by #4651. If you look at the GR file it's actually not so monolithic, it's actually quite easy to understand and way less code than two files for each function and then gluing them all together.

Comment (nilsvu) [reply src/PointwiseFunctions/Hydro/Python/Bindings.cpp:33]: Adding this to the docs: https://github.com/sxs-collaboration/spectre/pull/5385

**Comment by nilsvu** [comment]
Merged in #6186

**Pull request closed without merging**